src/gui/sdlmain.cpp

```diff
@@ -203,6 +208,9 @@ enum SCREEN_TYPES	{
 	SCREEN_OPENGL
 };
 
+/* Kinds of simple pixel scaling, currently encoded in the output type: */
+enum OutKind {OkNone, OkNearest, OkPerfect};
```

Comment:
Good idea! `ScalingHint` is indeed much better than `OutKind`, but maybe `ScalingType`, `ScalingKind`, or `ScalingMode` is better still? *hint* is too lateral and indirect. What does @krcroft think—a native speaker of English, I trow?

As to `enum class`, I should prefer a plain `enum` for the brevity of identifiers: `StNearest` vs. `ScalingType::Nearest`, even though it be less type-safe.

Reply:
The closer we can get to a conversation description, the better.   I think `Hint` works if we're allowed to disregard it in the code and change to something different (I'm not sure; do we?). 

ScalingChoice, implies the scaling algorithm comes from a user's choice, and not from any other internally determined logic. (We could differentiate all user-configurable enums with the adjective 'Choice' or synonym). 

As for the actual scaling algorithm names  (and I don't want us to start shedding any bikes here), I've seen Nvidia's PP described as: 

> performing nearest neighbor scaling of lower-resolution images to an integer multiple of their original resolution, producing a sharp, pixelated image

I've also see people use 'lossless' to describe it.

So I'm actually confused by the term nearest vs perfect. 

For me; I try to imagine sitting beside a newbie, and what questions I would ask them to help them decide. 

"Do you prefer your DOS games take up your screen's full vertical resolution, and would accept some pixel blurring as a trade-off? If yes.. choose X.

"Or would you prefer your games' pixels be absolutely undistorted and blur-free if it meant the game might be shown at a smaller size, and not take up the full vertical resolution, to achieve pixel-perfection? If so, choose Y."

So I see the enum names as answering the questions:  "what's your priority?" or "What are you optimizing?".   I'm not sure of what I would pick.  Just throwing out thoughts to spur more ideas.

Reply:
*kcroft*:

> The closer we can get to a conversation description, the better. I think
> `Hint` works if we're allowed to disregard it in the code and change to
> something different (I'm not sure; do we?).

That is exactly why I like *Hint* the least. The `enum` in question
shall represent the scaling part of the currently composite `output`
setting, e.g: *Perfect* for `texturepp`, *Nearest* for `openglnb`, and
*None* for `opengl`. It comes from the parsing of the `output` setting,
so *None* represents whatever default scaling the rendering mode has.

> `ScalingChoice`, implies the scaling algorithm comes from a user's choice, and
> not from any other internally determined logic. (We could differentiate all
> user-configurable enums with the adjective *Choice* or synonym).

Well, I think this name a limiting factor that brings in unnecessary
considerations. In my opinion, the name of an `enum` shall represent
what behavior in the code in *controls* rather than whence it comes.
This puts focus on behavior and adds flexibility, allowing `ScalingMode`
or `ScalingType` to originate from internal fall-back logic instead of
explicit user choice (for example).

> As for the actual scaling algorithm names (and I don't want us to start
> shedding any bikes here), I've seen Nvidia's PP described as:
> 
> > performing nearest neighbor scaling of lower-resolution images to an
> > integer multiple of their original resolution, producing a sharp, pixelated
> > image
> 
> I've also see people use 'lossless' to describe it.
> 
> So I'm actually confused by the term nearest vs perfect.

I will explain. Pixel-perfect scaling may be considered as a special
case of nearest-neighbor scaling, even as my current implementation
considers it, by implementing it via nearest-neighbor scaling
into an output window of precisly calculated dimensions.

Pixel-perfect scaling is lossless in the sense that it preserves
absolutely sharp pixels of regular size.  In interpolative scaling, the
pixels become blurry, whereas in the general case of nearest-neighbor
scaling they are irregular (of variying size). Imagine nearest-neighbor
upscaling of a linear array of size 4 into size 6 (by a non-integer
factor of 1.5):

    1 2 3 4 => 1 2 2 3 4 4

— the sizes of the resuliting elemets altenate between one and two to
yield the disired final length:

    1 + 2 + 1 + 2 = 6

> For me; I try to imagine sitting beside a newbie, and what questions I would
> ask them to help them decide.
> 
> "Do you prefer your DOS games take up your screen's full vertical resolution,
> and would accept some pixel blurring as a trade-off? If yes.. choose X.
> 
> "Or would you prefer your games' pixels be absolutely undistorted and
> blur-free if it meant the game might be shown at a smaller size, and not take
> up the full vertical resolution, to achieve pixel-perfection? If so, choose
> Y."

I think these would be great, user-friendly documentation for the config
file!

> So I see the enum names as answering the questions: "what's your
> priority?" or "What are you optimizing?".

Understood. I think it is too user-oriented for internal programming,
and readers familliar with the basics of raster images with feel
insulted by such overcondescending attitude. When one knows what one is
doing, layman's euphemitic circumventinve phrasing becomes much of an
annoyance, by conceling what is actually going on.

Technical terms are shorter, and more direct. But the documentation and
comments in the .INI-file should be as accessible to non-technical
people as possible. This is like a difference between a functional
specification and end-user documentation: both describe the behavior of
the system as a black box, but address different audiences.

> I don't have any terse words for these differing priorities yet..
> "DisplayPriority" vs "PixelPriority", or
> .. "FullStretch" vs."IntegerStretch", or ... none of these are very
> good.. but maybe spur more ideas. Those come at the problem from the
> user perspective; similar to a car's "sport mode" vs "economy mode".
>
> Another approach (favoured by engineers like myself, and users who
> like cool terms), is we "termify" the underlying technology. So
> instead of "sport mode", you name it "LGR-DS Turbo", standing for "low
> gear-ratio with delayed shift", and have dudes lining up to buy it :-)

I have nothing against *sports mode* because it is a short name not
supplementing, but comprising a set internal parameters.
It can even be more useful due to encapsulation: the driver may
fine-tune the parameters underlying *sports mode*.

> If we're descriptive enough (even with comparison screenshots on the
> wiki), I see both approaches working just fine.
> 
> And hey.. this feature is as good as it gets in terms of display
> correctness (like nuked for adlib), and IMO, warrants a name that
> carries some gravitas in Dosbox's core settings.

I should prefer to keep it modest, neutral, calm, and techinally
correct. Pixel-perfect scaling is not a big deal anyway, and CD playback
with on-the-fly resampling is magnitudes harder to implement.

Reply:
Something is wrong. *Markdown* here is not handling line breaks correctly. It should join them!

Reply:
I suggested word "hint" because so far it translated to `SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, …)` call (this is only one of many hints available in SDL2).

Reply:
Thanks for the explanation and reasoning @ant-222. 

I think we've got three areas of decisions:
- config options
- internal enum names
- user-side documentation

PP's existing config names, by way of ECE, are well known to users and this sets a good baseline and precident. Suggest we use them or an obvious/intuitive variation on them, like you've suggested.

Given the scaling algorithm is not adjustable and the code uses whatever the user has set, I suggest either scaling type or mode. 

I think the decision to use namespaced enums or not, such as `ScalingMode::Nearest`, `ScalingMode::Perfect` vs. `ScaleNearest`, `ScalePerfect` should:

1. relate to our config key and values -- so there's and obvious and simple connection between the two. 

 2. I think the decision should be made  convention that we can adhere to going forward, and atleast eliminate ambiguity for future changes and code reviews. 

As I'm not an expert in the underlying pros and cons, I will defer my judgement to what @dreamer recommends. 

Likewise, the modernish c++ references mostly prefer scoped/namespaced - so that will get my default vote (because this has no doubt undergone far more extensive discussion by experts with wisdom far beyond myself).

As for documentation: I think we're already concluded that user-focused verbose help should be provided (in the config and readme). 

What do you think?  Please weigh in on the 'scale'  ⚖️ :-)

Reply:
> PP's existing config names, by way of ECE, are well known to users and this sets a good baseline and precident. Suggest we use them as a baseline with minor adjustments to meet overall project direction.

Even *I* think that *pp* it too cryptic. When we separate *renderer* and *scaling*, the scaling settings should have a more sensible values than *pp*, *nb*, and *np*. Let us consider `pixel-perfect`, `nearest-neighbor`, and `near-perfect` for a start.

> Given the scaling algorithm is not adjustable and the code uses whatever the user has set, I suggest either scaling type or mode.

Well, it has one adjustable parameter—`aspweight`—that is currently hard-coded. It determines the weight of aspect ratio in the calculation of the best scaling factors.

I like both `type` and `mode`, and `mode` a tad better than `type`. I dislike `hint` and `kind`.
`hint` makes an impression that it is optional, not mandatory. One is not forced to follow a hint. `kind` is too rare in such contexts.

> I think the decision to use namespaced enums or not, such as `ScalingMode::Nearest`, `ScalingMode::Perfect` vs. `ScaleNearest`, `ScalePerfect` should:
> 
> 1. relate to our config key and values -- so there's and obvious and simple connection between the two.
> 
> 2. become a project-wide convention that we can adhere to going forward.  The reason being that we want to use best practices and also eliminate ambiguity for future changes and code reviews like this (especially code that involves connected config key/vals and enums in the code).
>        
> To nail this down, suggest a short description and example be added to the convention wiki page. I'm happy to make this wiki revision, once we conclude here.
>        
> As I'm not an expert in the underlying pros and cons, I will defer my judgement to what @dreamer recommends.
>        
> Likewise, the modernish c++ references mostly prefer scoped/namespaced - so that will get my default vote (because this has no doubt undergone far more extensive discussion by experts with wisdom far beyond myself).

I cannot add anything to what I have already said. I find the `::` syntax right ugly and qualified enums too cumbersome. In *C*, I prefer the following approach:
~~~
enum ScalingMode {ScPixelPerfect, ScNearPerfect, ScNearest, ScNone}
~~~
but the verbosity of *C++* tradition may require identifers half-screen long, e.g. `ScalingMode::NearestNeighbor` takes 30 characters, which is 38% of line width.

> documentation: I think we're in agreement that user-focused verbose help should be provided in the config and readme.

Yes, but it should take care not to make the reader feel stupid, or to circumlocute in the layman's language at the expense of precision. I suggest technical terminology with liberal comments and annotations.

Reply:
> Let us consider pixel-perfect, nearest-neighbor, and near-perfect for a start.

These are great! 

One thing that hit my was this hypothetical situation: _"Good day sir, are you interested in buying a BBQ? We have the perfect Weber-XT unit here and sitting out back we have the near-perfect Jameson 200.  Both cost the same."_   Although I'm not sure how to adjust "near-perfect" wording. 

I agree `ScalingMode::NearestNeighbor` and `ScalingMode::PixelPerfect` are longer than your coding style, but fortuntaely these are likely only used in conditionals during config parsing or stearing code - so they will likely be alone on their own line:

``` c++
if (scaling_choice == "pixel-perfect")
    sm = ScalingMode::PixelPerfect;

else if (scaling_choice == "nearest-neighbor")
    sm = ScalingMode::NearestNeighbour;
```

Or

``` c++
if (sm == ScalingMode::PixelPerfect) {
   scaling_multiplier = perform_pp_calc( .. )
   pixel_aspect = do_something_else(..)
} else if (sm == ScalingMode::NearestNeighbor) {
   scaling_multiplier = perform_nn_calc( .. )
   pixel_aspect = do_something_else(..)
}
```

but If you were doing calculations with them (and all togeher), then yeah - it would get pretty messy!

Namespaces can certainly get far more bloated than this. I've seen nasty stuff like:
 `apiextensions::spec::validation::openAPIV3Schema::properties::replicas::type`.

Just imagine something like: `DOSboxAPIv2::UserSettings::SDL::Screen::ScalingMode::PixelPerfect`.. :-)

Reply:
> I suggested word "hint" because so far it translated to `SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, …)` call (this is only one of many hints available in SDL2).

I think *SDL* has more reason to call them hits, because they are not necessarily [obeyed](https://wiki.libsdl.org/SDL_SetHint). I fear it will be a false analogy if we identify scaling algorithms via an `enum` named `ScalingHint`...

Reply:
> I think _SDL_ has more reason to call them hits, because they are not necessarily [obeyed](https://wiki.libsdl.org/SDL_SetHint). I fear it will be a false analogy if we identify scaling algorithms via an `enum` named `ScalingHint`...

Good point. `ScalingMode` is better name.